### ISSUE TITLE
lxc stop / lxc restart are not reliable and sometimes hang

### ISSUE THREAD

**Issue by tchwpkgorg** [lxc/lxd#1752]
lxc stop / lxc restart are not reliable and sometimes hang.

I'm stopping / restarting large number of containers every day, and have noticed that sometimes, they just hang without any apparent reason.

The reproducer is a bit extreme (typically, no one would stop/start/restart the containers in such patterns), but it's just to make reproducing easier (the issue was happening for me with "normal" usage and happens randomly, especially if the host is IO-bound).

1) reproducer for "lxc stop" hang:

CONTAINER=containername
while true; do echo start; time lxc start $CONTAINER; echo stop; time lxc stop $CONTAINER;  done

You will notice it hangs at "stop" after a few iterations

2) reproducer for "lxc restart" hang:

CONTAINER=containername
while true; do echo restart; time lxc restart $CONTAINER; done

You will notice if hangs at "restart" after a few iterations.

This is somehow similar to https://github.com/lxc/lxd/issues/1731

Ubuntu 14.04 with 4.4.5 kernel:

```
ii  lxd                                      2.0.0~rc3-0ubuntu1~ubuntu14.04.1~ppa1  amd64        Container hypervisor based on LXC - daemon
ii  lxd-client                               2.0.0~rc3-0ubuntu1~ubuntu14.04.1~ppa1  amd64        Container hypervisor based on LXC - client
ii  lxd-tools                                2.0.0~rc3-0ubuntu1~ubuntu14.04.1~ppa1  amd64        Container hypervisor based on LXC - extra tools
```

**Comment by stgraber** [comment]
Unless you can reproduce this issue when passing "--force", it is not a LXD issue.

Both stop and restart without --force only send a signal to PID 1 in the container, SIGPWR for stop and SIGINT for restart. If the signal handler isn't in place yet or if PID 1 ignores or fails to perform the required action, the command will wait indefinitely.

You can specify a timeout for either command, after which you'll get a failure message if you prefer that behavior. We don't set a timeout by default as while most containers will shutdown or reboot within 30s, it is not unusual for database servers to take several minutes in which case forcing the action would cause data loss.

**Comment by tchwpkgorg** [comment]
Can we at least expect an equivalent of the following commands from lxc-stop?

```
  -W, --nowait      don't wait for shutdown or reboot to complete
  -t, --timeout=T   wait T seconds before hard-stopping
```

**Comment by tchwpkgorg** [comment]
I have it hanging with --force:

```
# lxc stop z-testing-a19ea622182c63ddc19bb22cde982b82-2016-03-14-06-57-28 --force --debug
DBUG[03-14|14:47:27] Raw response: {"type":"sync","status":"Success","status_code":200,"metadata":{"api_extensions":[],"api_status":"development","api_version":"1.0","auth":"trusted","config":{"core.https_address":"10.190.0.1:8443","core.trust_password":true},"environment":{"addresses":["10.190.0.1:8443"],"architectures":["x86_64","i686"],"certificate":"-----BEGIN CERTIFICATE-----\n(...)\n-----END CERTIFICATE-----\n","driver":"lxc","driver_version":"2.0.0.rc10","kernel":"Linux","kernel_architecture":"x86_64","kernel_version":"4.4.5-040405-generic","server":"lxd","server_pid":32402,"server_version":"2.0.0.rc3","storage":"btrfs","storage_version":"4.4"},"public":false}}

DBUG[03-14|14:47:27] Raw response: {"type":"sync","status":"Success","status_code":200,"metadata":{"architecture":"x86_64","config":{"raw.lxc":"lxc.aa_allow_incomplete=1","volatile.base_image":"1032903165a677e18ed93bde5057ae6287841ae756d1a6296eef8f2e5a825e4a","volatile.eth0.hwaddr":"00:16:3e:f4:d3:9f","volatile.eth0.name":"eth0","volatile.last_state.idmap":"[{\"Isuid\":true,\"Isgid\":false,\"Hostid\":100000,\"Nsid\":0,\"Maprange\":65536},{\"Isuid\":false,\"Isgid\":true,\"Hostid\":100000,\"Nsid\":0,\"Maprange\":65536}]"},"created_at":"2016-03-14T06:57:31Z","devices":{"eth0":{"nictype":"bridged","parent":"br-testing","type":"nic"},"root":{"path":"/","type":"disk"},"uploads":{"path":"/var/www/uploads","source":"/srv/deployment/uploads","type":"disk"}},"ephemeral":false,"expanded_config":{"raw.lxc":"lxc.aa_allow_incomplete=1","volatile.base_image":"1032903165a677e18ed93bde5057ae6287841ae756d1a6296eef8f2e5a825e4a","volatile.eth0.hwaddr":"00:16:3e:f4:d3:9f","volatile.eth0.name":"eth0","volatile.last_state.idmap":"[{\"Isuid\":true,\"Isgid\":false,\"Hostid\":100000,\"Nsid\":0,\"Maprange\":65536},{\"Isuid\":false,\"Isgid\":true,\"Hostid\":100000,\"Nsid\":0,\"Maprange\":65536}]"},"expanded_devices":{"eth0":{"nictype":"bridged","parent":"br-testing","type":"nic"},"root":{"path":"/","type":"disk"},"uploads":{"path":"/var/www/uploads","source":"/srv/deployment/uploads","type":"disk"}},"name":"z-testing-a19ea622182c63ddc19bb22cde982b82-2016-03-14-06-57-28","profiles":["default"],"stateful":false,"status":"Running","status_code":103}}

DBUG[03-14|14:47:27] Putting {"action":"stop","force":true,"stateful":false,"timeout":-1}
 to http://unix.socket/1.0/containers/z-testing-a19ea622182c63ddc19bb22cde982b82-2016-03-14-06-57-28/state
DBUG[03-14|14:47:27] Raw response: {"type":"async","status":"Operation created","status_code":100,"metadata":{"id":"fd9a941c-f071-43b0-8a4d-35edefae5ef3","class":"task","created_at":"2016-03-14T14:47:27.433093876Z","updated_at":"2016-03-14T14:47:27.433093876Z","status":"Running","status_code":103,"resources":{"containers":["/1.0/containers/z-testing-a19ea622182c63ddc19bb22cde982b82-2016-03-14-06-57-28"]},"metadata":null,"may_cancel":false,"err":""},"operation":"/1.0/operations/fd9a941c-f071-43b0-8a4d-35edefae5ef3"}

DBUG[03-14|14:47:27] 1.0/operations/fd9a941c-f071-43b0-8a4d-35edefae5ef3/wait


```


In the meantime, I can get into the container and list the running processes:

```
# lxc exec z-testing-a19ea622182c63ddc19bb22cde982b82-2016-03-14-06-57-28 /bin/bash
root@z-testing-a19ea622182c63ddc19bb22cde982b82-2016-03-14-06-57-28:~# ps auxf
USER       PID %CPU %MEM    VSZ   RSS TTY      STAT START   TIME COMMAND
root      5348  0.0  0.0  18192  3320 ?        Ss   14:49   0:00 /bin/bash
root      5358  0.0  0.0  15568  2220 ?        R+   14:49   0:00  \_ ps auxf
root         1  0.0  0.0  33388  2892 ?        Ss   06:57   0:00 /sbin/init
root      1495  0.0  0.0  15264  1188 ?        S    06:57   0:00 upstart-socket-bridge --daemon
root      1697  0.0  0.0  19480     0 ?        S    06:57   0:00 upstart-udev-bridge --daemon
root      1701  0.0  0.0  49276  1824 ?        Ss   06:57   0:00 /lib/systemd/systemd-udevd --daemon
message+  1837  0.0  0.0  39224  1296 ?        Ss   06:57   0:00 dbus-daemon --system --fork
root      1862  0.0  0.0  15280     4 ?        S    06:57   0:00 upstart-file-bridge --daemon
root      1870  0.0  0.0  35024  1696 ?        Ss   06:57   0:00 /lib/systemd/systemd-logind
daemon    1901  0.0  0.0  19144    16 ?        Ss   06:57   0:00 atd
root      1902  0.0  0.0  23660  1924 ?        Ss   06:57   0:00 cron
root      1905  0.0  0.0   4372  1456 ?        Ss   06:57   0:00 acpid -c /etc/acpi/events -s /var/run/acpid.socket
syslog    1932  0.0  0.0 182116  1904 ?        Ssl  06:57   0:00 rsyslogd
root      2015  0.0  0.0  19196  1532 ?        Ss   06:57   0:03 /usr/sbin/irqbalance
root      2061  0.0  0.0  25348  2608 ?        Ss   06:57   0:00 /usr/lib/postfix/master
postfix   2064  0.0  0.0  27576  2680 ?        S    06:57   0:00  \_ qmgr -l -t unix -u
postfix   4859  0.0  0.0  40184  3044 ?        S    07:01   0:00  \_ tlsmgr -l -t unix -u -c
postfix   5273  0.0  0.0  27412  2812 ?        S    13:38   0:00  \_ pickup -l -t unix -u -c
vnstat    2083  0.0  0.0   7480  1516 ?        Ss   06:57   0:01 /usr/sbin/vnstatd -d --pidfile /run/vnstat/vnstat.pid
root      2124  0.0  0.0  61384  2072 ?        Ss   06:57   0:00 /usr/sbin/sshd -D
root      2151  0.0  0.0 397432 17496 ?        Ss   06:57   0:03 php-fpm: master process (/etc/php5/fpm/php-fpm.conf)
www-data  2162  0.0  0.1 414508 56112 ?        S    06:57   0:00  \_ php-fpm: pool www
www-data  2163  0.0  0.2 414388 67028 ?        S    06:57   0:00  \_ php-fpm: pool www
www-data  2164  0.0  0.2 420988 72864 ?        S    06:57   0:00  \_ php-fpm: pool www
www-data  2165  0.0  0.1 415840 65356 ?        S    06:57   0:00  \_ php-fpm: pool www
www-data  2166  0.0  0.1 406580 33492 ?        S    06:57   0:00  \_ php-fpm: pool www
www-data  2167  0.0  0.0 404928 30444 ?        S    06:57   0:00  \_ php-fpm: pool www
www-data  2168  0.0  0.0 404156 24984 ?        S    06:57   0:00  \_ php-fpm: pool www
www-data  2169  0.0  0.0 399628 23016 ?        S    06:57   0:00  \_ php-fpm: pool www
www-data  2170  0.0  0.2 419136 72940 ?        S    06:57   0:01  \_ php-fpm: pool www
www-data  2171  0.0  0.1 408136 44508 ?        S    06:57   0:00  \_ php-fpm: pool www
www-data  2172  0.0  0.0 403908 25764 ?        S    06:57   0:00  \_ php-fpm: pool www
www-data  2173  0.0  0.0 399588 22568 ?        S    06:57   0:00  \_ php-fpm: pool www
www-data  2174  0.0  0.1 403116 45452 ?        S    06:57   0:00  \_ php-fpm: pool www
www-data  2175  0.0  0.1 406964 48412 ?        S    06:57   0:00  \_ php-fpm: pool www
www-data  2176  0.0  0.1 407468 62416 ?        S    06:57   0:01  \_ php-fpm: pool www
www-data  2177  0.0  0.1 406740 53120 ?        S    06:57   0:00  \_ php-fpm: pool www
www-data  2178  0.0  0.1 406248 57672 ?        S    06:57   0:00  \_ php-fpm: pool www
www-data  2179  0.0  0.1 406460 51840 ?        S    06:57   0:00  \_ php-fpm: pool www
www-data  2180  0.0  0.1 408684 48568 ?        S    06:57   0:00  \_ php-fpm: pool www
www-data  2181  0.0  0.1 404392 46060 ?        S    06:57   0:00  \_ php-fpm: pool www
root      4807  0.0  0.0  32492  1540 ?        Ss   07:00   0:00 nginx: master process /usr/sbin/nginx -c /etc/nginx/nginx.conf
www-data  4808  0.0  0.0  32800  3068 ?        S    07:00   0:01  \_ nginx: worker process
www-data  4809  0.0  0.0  32496  1812 ?        S    07:00   0:01  \_ nginx: worker process
www-data  4810  0.0  0.0  32652  2800 ?        S    07:00   0:00  \_ nginx: worker process
www-data  4811  0.0  0.0  32496  1812 ?        S    07:00   0:01  \_ nginx: worker process

```

**Comment by hallyn** [comment]
So this is the same container causing troublein ##1753.  Would be useful to
know how to get the container monitor into that state reliably (maybe just
sig-stopping the container would work) and then make sure lxc can work
sanely in all respects with such containers around.

**Comment by stgraber** [comment]
@hallyn can I assign this one to you?

**Comment by hallyn** [comment]
> @hallyn can I assign this one to you?

Sure.

**Comment by hallyn** [comment]
@tchwpkgorg could you show a 'ps -ef' output when one of the commands is stopped?  I have posted a PR to work around the hang, but would like to know what's causing the actual hang.  (Which is likely a lxc bug).

**Comment by hallyn** [comment]
I've had a start stop loop like you showed at top going for 1.5 hrs with no hang.  Can you show how you created the container in the first place, as well as your lxd and lxc versions?  (and still ps -ef output)

**Comment by stgraber** [comment]
Keeping this one open as it'd be nice to root cause it before closing since the pull request is just a workaround to prevent such issues from hanging lxd entirely.

**Comment by stgraber** [comment]
Removing the milestone tag though as it's unclear what's wrong and the workaround has been merged.

**Comment by srkunze** [comment]
Reminds me of #1336.

**Comment by hallyn** [comment]
No, #1336 is the container init not responding to a shutdown signal.  This is about the container monitor, lxc itself, not responding.

**Comment by srkunze** [comment]
@hallyn I see. Thanks for clarifying.

**Comment by stgraber** [comment]
Tentatively closing this. It seems like the issue was a hung monitor process for some reason, at least now we have proper reporting of such issue so let us know if it happens again and we'll definitely look into it.

**Comment by saulcosta18** [comment]
I'm running into what I believe to be this issue on Ubuntu 16.04 (with Ubuntu 16.04 containers). Essentially, start / stop appear to work, but do not complete (i.e. I have to CTRL+C out of them, but `lxc list` shows the command actually did complete).

From reading this and other issues, it looks like `lxc-monitord` may be the culprit, and `strace` does reveal that it's stuck.

Higher loads (hundreds of containers on a host) appears to aggravate the issue.

Do you have any recommendations for next steps in debugging this?

**Comment by brauner** [comment]
So I believe that I have identified the underlying cause/bug in `LXC`. The monitor related fixes should will be in `LXC 2.0.7`: https://github.com/lxc/lxc/pull/1327, https://github.com/lxc/lxc/pull/1328, https://github.com/lxc/lxc/pull/1336, https://github.com/lxc/lxc/pull/1331. They all fix a bunch of possible deadlocks. The one most likely causing your issue seems to be https://github.com/lxc/lxc/pull/1327.
